src/hotspot/share/classfile/loaderConstraints.cpp

```diff
@@ -185,62 +191,54 @@ class PurgeUnloadedConstraints : public StackObj {
     LogTarget(Info, class, loader, constraints) lt;
     int len = set.num_constraints();
     for (int i = len - 1; i >= 0; i--) {
-      LoaderConstraint* probe = set.constraint_at(i);
-      InstanceKlass* klass = probe->klass();
-      // Remove klass that is no longer alive
-      if (klass != NULL &&
-          !klass->is_loader_alive()) {
-        probe->set_klass(NULL);
+      LoaderConstraint* constraint = set.constraint_at(i);
+      InstanceKlass* klass = constraint->klass();
+
+      // Remove klass and constraint that is no longer alive
+      if (klass != NULL && !klass->is_loader_alive()) {
         if (lt.is_enabled()) {
           ResourceMark rm;
           lt.print("purging class object from constraint for name %s,"
                      " loader list:",
                      name->as_C_string());
-          for (int i = 0; i < probe->num_loaders(); i++) {
+          for (int i = 0; i < constraint->num_loaders(); i++) {
             lt.print("    [%d]: %s", i,
-                          probe->loader_data(i)->loader_name_and_id());
+                          constraint->loader_data(i)->loader_name_and_id());
           }
         }
-      }
-
-      // Remove entries no longer alive from loader array
-      for (int n = probe->num_loaders() - 1; n >= 0; n--) {
-        if (probe->loader_data(n)->is_unloading()) {
-          if (lt.is_enabled()) {
-            ResourceMark rm;
-            lt.print("purging loader %s from constraint for name %s",
-                     probe->loader_data(n)->loader_name_and_id(),
-                     name->as_C_string());
-          }
-          probe->remove_loader_at(n);
-
-          if (lt.is_enabled()) {
-            ResourceMark rm;
-            lt.print("new loader list:");
-            for (int i = 0; i < probe->num_loaders(); i++) {
-              lt.print("    [%d]: %s", i,
-                            probe->loader_data(i)->loader_name_and_id());
+        set.remove_constraint(name, constraint);
+      } else {
+        // The class is alive or null but some of of the loaders may not be.
+        // Remove entries no longer alive from loader array
+        for (int n = constraint->num_loaders() - 1; n >= 0; n--) {
```

Comment:
I'm curious why we iterate backwards here?

Reply:
We iterate backwards so that I can remove the entries at the end first and can still use indexing to loop through the growable array.  We do that with growable arrays in a lot of places.